Fix getting section info in large mach-o files.

clayborg opened the pull request:
Mach-o has 32 bit file offsets in the MachO::section_64 structs. dSYM files can contain sections whose start offset exceeds UINT32_MAX, which means the MachO::section_64.offset will get truncated. We can calculate when this happens and properly adjust the section offset to be 64 bit safe. This means tools can get the correct section contents for large dSYM files and allows tools that parse DWARF, like llvm-gsymutil, to be able to load and convert these files correctly.

clayborg committed:
Fix getting section info in large mach-o files.

Mach-o has 32 bit file offsets in the MachO::section_64 structs. dSYM files can contain sections whose start offset exceeds UINT32_MAX, which means the MachO::section_64.offset will get truncated. We can calculate when this happens and properly adjust the section offset to be 64 bit safe. This means tools can get the correct section contents for large dSYM files and allows tools that parse DWARF, like llvm-gsymutil, to be able to load and convert these files correctly.

DataCorrupted commented on llvm/lib/Object/MachOObjectFile.cpp:
```diff
@@ -1978,20 +1978,34 @@ uint64_t MachOObjectFile::getSectionSize(DataRefImpl Sec) const {
+          (uint64_t)CurrSect.offset + CurrSect.size;
+      if (EndSectFileOffset >= UINT32_MAX)
+        SectOffsetAdjust += EndSectFileOffset & 0xFFFFFFFF00000000ull;
+    }
```
Thanks for the change, I'm all for it. Maybe we should loop some apple folks in as well.

Minor comments on the code:
We are building this adjustment on a strong assumption that section offsets are ordered, this might not be true for MachO (although most likely). 

Assert and bail out if sections are not ordered in case we run into any corner cases that are not dSYM.

```suggestion
    // adjust the section offset to be 64 bit safe.
    // Assumes the sections are ordered.
    uint64_t PrevTrueOffset = 0;
    uint64_t SectOffsetAdjust = 0;
    for (uint32_t SectIdx=0; SectIdx<Sec.d.a; ++SectIdx) {
      MachO::section_64 CurrSect =
          getStruct<MachO::section_64>(*this, Sections[SectIdx]);
       uint64_t CurrTrueOffset = (uint64_t) CurrSect.offset | SectOffsetAdjust;
      assert(SectOffsetAdjust == 0 || (PrevTrueOffset <= CurrTrueOffset) && "Overflowing sections must be ordered for adjustment")
      const uint64_t EndSectFileOffset =
          (uint64_t)CurrSect.offset + CurrSect.size;
      if (EndSectFileOffset >= UINT32_MAX)
        SectOffsetAdjust += EndSectFileOffset & 0xFFFFFFFF00000000ull;
      PrevTrueOffset = CurrTrueOffset;
    }
```

clayborg committed:
Address user feedback to check for increasing sections sizes.

We now return an error if a section file offset exceeds 4GB and the sections are not ordered in the mach-o file. If sections are not ordered, we can't assume the section file offset overflows make sense to apply to other sections, but we can if they are ordered.

clayborg committed:
Fix comment to reflect new code.

clayborg committed:
Run clang format.

clayborg committed:
Improve error message when section data exceeds 4GB.

clayborg commented:
> > Thanks for fixing! Can we add a test?
> 
> Just tested locally, some 8GB dSYM was consumed successfully by llvm-dwarfdump and gsymutil. I don't have any out-of-order larger-than-4GB dSYM to test the error handling, and I think that's fairly rare case. In terms of checking a test file into the git, I think our experience in lldb (#164471) told us that its just hard to do.

We definitely can't commit a huge binary, and obj2yaml amd yaml2obj won't create invalid object files (where the segments would have the numbers for us that we need without having the file data), nor do we want them to emit 4 to 8 GB binaries to disk just so we can test them. I also didn't see a Test.cpp variant for llvm/lib/Object/MachOObjectFile.cpp anywhere. Also if you create a mach-o file with just some LC_SEGMENT_64 load commands, but the data for the file isn't in the file, it will not load and emit an error I believe. So this is hard to test. I am open to ideas if anyone has any, but I can't find any acceptable solutions.
